### PR TITLE
[tsan] Don't use `enum __tsan_memory_order` in tsan interface

### PR DESCRIPTION
**Pull request by vitalybuka**
In C++ it's UB to use undeclared values as enum.
And there is support `__ATOMIC_HLE_ACQUIRE` and
`__ATOMIC_HLE_RELEASE` need such values.

Internal implementation was switched to `class enum`,
where that behavior is defined. But interface is C, so
we just switch to `int`.

### CODE REVIEW

File: compiler-rt/lib/tsan/rtl/tsan_interface.h [hunk 1]
```diff
@@ -219,14 +219,14 @@ __extension__ typedef __int128 a128;
 
 // Part of ABI, do not change.
```
Comment (melver) [inline compiler-rt/lib/tsan/rtl/tsan_interface.h:220]: Is it still following the ABI - if we do not static_cast it to int, then I don't think this is strictly required to follow ABI.

**Review by melver — APPROVED** [review]

File: compiler-rt/lib/tsan/rtl/tsan_interface.h (hunk 1 above)
Comment (vitalybuka) [reply compiler-rt/lib/tsan/rtl/tsan_interface.h:220]: Does #114916 help?

**Review by melver — APPROVED** [review]

**Comment by felipepiovezan** [comment]
This is breaking the greendragon bots, could you revert or submit a fix if it's a quick one?

```
/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:103:1: error: no matching function for call to '__tsan_atomic32_fetch_add'
  103 | OSATOMIC_INTERCEPTORS_ARITHMETIC(OSAtomicAdd, fetch_add,
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  104 |                                  OSATOMIC_INTERCEPTOR_PLUS_X)
      |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:81:37: note: expanded from macro 'OSATOMIC_INTERCEPTORS_ARITHMETIC'
   81 |     m(int32_t, int32_t, a32, f##32, __tsan_atomic32_##tsan_atomic_f,         \
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:6:1: note: expanded from here
    6 | __tsan_atomic32_fetch_add
      | ^~~~~~~~~~~~~~~~~~~~~~~~~
/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:63:14: note: expanded from macro 'OSATOMIC_INTERCEPTOR_PLUS_X'
   63 |       return tsan_atomic_f((volatile tsan_t *)ptr, x, mo) + x;               \
      |              ^~~~~~~~~~~~~
/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interface.h:278:5: note: candidate function not viable: no known conversion from 'const morder' to 'int' for 3rd argument
  278 | a32 __tsan_atomic32_fetch_add(volatile a32 *a, a32 v, int mo);
      |     ^                                                 ~~~~~~
```
https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/14686/console

**Comment by vitalybuka** [comment]
This is breaking the greendragon bots, could you revert or submit a fix if it's a quick one?

> /Users/ec2-user/jenkins/workspace/[llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:103:1](http://llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:103:1): error: no matching function for call to '__tsan_atomic32_fetch_add' 103 | OSATOMIC_INTERCEPTORS_ARITHMETIC(OSAtomicAdd, fetch_add, | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 104 | OSATOMIC_INTERCEPTOR_PLUS_X) | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ /Users/ec2-user/jenkins/workspace/[llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:81:37](http://llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:81:37): note: expanded from macro 'OSATOMIC_INTERCEPTORS_ARITHMETIC' 81 | m(int32_t, int32_t, a32, f##32, __tsan_atomic32_##tsan_atomic_f, | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ :6:1: note: expanded from here 6 | __tsan_atomic32_fetch_add | ^~~~~~~~~~~~~~~~~~~~~~~~~ /Users/ec2-user/jenkins/workspace/[llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:63:14](http://llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:63:14): note: expanded from macro 'OSATOMIC_INTERCEPTOR_PLUS_X' 63 | return tsan_atomic_f((volatile tsan_t *)ptr, x, mo) + x; | ^~~~~~~~~~~~~ /Users/ec2-user/jenkins/workspace/[llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interface.h:278:5](http://llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interface.h:278:5): note: candidate function not viable: no known conversion from 'const morder' to 'int' for 3rd argument 278 | a32 __tsan_atomic32_fetch_add(volatile a32 *a, a32 v, int mo); | ^ ~~~~~~ https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/14686/console

**Comment by vitalybuka** [comment]
This is breaking the greendragon bots, could you revert or submit a fix if it's a quick one?

> /Users/ec2-user/jenkins/workspace/[llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:103:1](http://llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:103:1): error: no matching function for call to '__tsan_atomic32_fetch_add' 103 | OSATOMIC_INTERCEPTORS_ARITHMETIC(OSAtomicAdd, fetch_add, | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 104 | OSATOMIC_INTERCEPTOR_PLUS_X) | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ /Users/ec2-user/jenkins/workspace/[llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:81:37](http://llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:81:37): note: expanded from macro 'OSATOMIC_INTERCEPTORS_ARITHMETIC' 81 | m(int32_t, int32_t, a32, f##32, __tsan_atomic32_##tsan_atomic_f, | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ :6:1: note: expanded from here 6 | __tsan_atomic32_fetch_add | ^~~~~~~~~~~~~~~~~~~~~~~~~ /Users/ec2-user/jenkins/workspace/[llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:63:14](http://llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_mac.cpp:63:14): note: expanded from macro 'OSATOMIC_INTERCEPTOR_PLUS_X' 63 | return tsan_atomic_f((volatile tsan_t *)ptr, x, mo) + x; | ^~~~~~~~~~~~~ /Users/ec2-user/jenkins/workspace/[llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interface.h:278:5](http://llvm.org/as-lldb-cmake/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interface.h:278:5): note: candidate function not viable: no known conversion from 'const morder' to 'int' for 3rd argument 278 | a32 __tsan_atomic32_fetch_add(volatile a32 *a, a32 v, int mo); | ^ ~~~~~~ https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/14686/console

**Pull request merged**